### PR TITLE
ZNTAMT-14 Create User table and concept of admin users

### PR DESCRIPTION
**Pull request by huangp**
https://zanata.atlassian.net/browse/ZNTAMT-14

when MT starts, it will print an initial password to the log console and write to a file ~~$HOME/initialPassword~~ $HOME/magpie_initial_password.

REST endpoint to create accounts
```
POST /api/account
// payload
{
	"name": "John Smith",
	"email": "jsmith@example.com",
	"accountType": "Normal",
	"roles": ["admin"],
	"credentials": [{
		"username": "devID",
	"secret": ["d", "e", "v", "K", "E", "Y"]
	}
		]
	
}
```
Once an account is created, the initial password is no longer valid.

Another REST endpoint to query all accounts (admin only)
```
GET /api/account
```

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-mt/53)

### CODE REVIEW

**Comment by seanf** [comment]
@huangp codecov says the coverage is going down.

**Comment by seanf** [comment]
Reviewed 29 of 46 files at r1, 4 of 6 files at r2, 14 of 14 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java, line 87 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zgQWa_lhWJFphDD7X:-L3zgQWa_lhWJFphDD7Y:b-ewyb8z) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java#L87)):*
> ```Java
>             return accountService.authenticate(username, token);
>         } else if (Objects.equals("admin", username)
>                 && Objects.equals(token, initialPass)) {
> ```

Given that "admin" and initialPass are not null, there's no need for Objects.equals here.

---

*[server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt, line 58 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3ze2r-ZouHh2_gRFbH:-L3ze2r-ZouHh2_gRFbI:bfhju7w) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt#L58)):*
> ```Kotlin
>         Mockito.verify(requestContext).abortWith(responseCaptor.capture())
> 
>         Assertions.assertThat(responseCaptor.value.status).isEqualTo(Response.Status.UNAUTHORIZED.statusCode)
> ```

Why aren't you importing the method `assertThat`?

---

*[server/src/test/java/org/zanata/magpie/integration/EnsureAdminAccountRule.kt, line 74 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zfIeoVGWjQxYwKDsQ:-L3zfIeoVGWjQxYwKDsR:blc2ffa) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/integration/EnsureAdminAccountRule.kt#L74)):*
> ```Kotlin
>     }
> 
>     private fun readInitialPasswordFromLog(): String {
> ```

...FromFile, I think.

---

*[server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt, line 65 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zY4ZlWCGt8-z5V1uY:-L3zY4Zm6ua_GsuOqeG-:b-paggae) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt#L65)):*
> ```Kotlin
> 
>         Assertions.assertThat(violation).hasSize(1)
>         Assertions.assertThat(violation.iterator().next().reason).isEqualTo("You don't have the necessary access")
> ```

Doesn't AssertJ have direct support for Sets?

I have used https://github.com/wuan/assertj-core-kotlin in zanata-platform, but these seem to be popular options too: https://github.com/kotlintest/kotlintest and https://markusamshove.github.io/Kluent/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by huangp** [comment]
Review status: 41 of 53 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java, line 87 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zgQWa_lhWJFphDD7X:-L4J-gSjFOGfwkg05oi8:b-896fix) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java#L87)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Given that "admin" and initialPass are not null, there's no need for Objects.equals here.
</blockquote></details>

Done.

---

*[server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt, line 58 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3ze2r-ZouHh2_gRFbH:-L4J-ivnCYpcMRLssCrb:b-rd5s42) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt#L58)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Why aren't you importing the method `assertThat`?
</blockquote></details>

I created an IntelliJ live template to output all the assert statements. It can only auto import packages not static import. 

---

*[server/src/test/java/org/zanata/magpie/integration/EnsureAdminAccountRule.kt, line 74 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zfIeoVGWjQxYwKDsQ:-L4J0L5luJ43_FJSfyOh:b-896fix) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/integration/EnsureAdminAccountRule.kt#L74)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

...FromFile, I think.
</blockquote></details>

Done.

---

*[server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt, line 65 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zY4ZlWCGt8-z5V1uY:-L4J0ufpVgVKVNWE1lQW:bmwh0kd) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt#L65)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Doesn't AssertJ have direct support for Sets?

I have used https://github.com/wuan/assertj-core-kotlin in zanata-platform, but these seem to be popular options too: https://github.com/kotlintest/kotlintest and https://markusamshove.github.io/Kluent/
</blockquote></details>

I need to get the element out and assert on the property "reason". The direct support for sets are all dealing with the element only.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by seanf** [comment]
Reviewed 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java, line 87 at r4](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L4JHjVXO0_X5EjdWHG7:-L4JHjVXO0_X5EjdWHG8:b-ohkcw6) ([raw file](https://github.com/zanata/zanata-mt/blob/765963cf6d731a98b2f6ff6140b53e83d5051a04/server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java#L87)):*
> ```Java
>             return accountService.authenticate(username, token);
>         } else if ("admin".equals(username)
>                 && token.equals(initialPass)) {
> ```

Do we need to make it `initialPass.equals(token)` to be safe?

---

*[server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt, line 58 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3ze2r-ZouHh2_gRFbH:-L4JI8qRKgbbvoD5bHsM:b-ylzt1m) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt#L58)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I created an IntelliJ live template to output all the assert statements. It can only auto import packages not static import. 
</blockquote></details>

That's a shame, because it's kind of bad for test readability. I don't suppose this helps: https://stackoverflow.com/a/22568349/14379

---

*[server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt, line 65 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zY4ZlWCGt8-z5V1uY:-L4JJcaKyqh3gHx5gCAi:by25y8d) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt#L65)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I need to get the element out and assert on the property "reason". The direct support for sets are all dealing with the element only.
</blockquote></details>

assertj (and thus assertj-core-kotlin) can do things like this:

    assertThat(fellowshipOfTheRing).extracting(TolkienCharacter::getName)
                               .doesNotContain("Sauron", "Elrond");

I think one of the other libraries can do something similar with this feature: https://github.com/kotlintest/kotlintest/blob/master/doc/reference.md#inspectors 

(KotlinTest looks to be quite powerful with its DSL, but as we're used to assertj perhaps we should stick with it for the moment.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by huangp** [comment]
Review status: 51 of 53 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java, line 87 at r4](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L4JHjVXO0_X5EjdWHG7:-L4JZRxSOkDXWGtxW_lb:bl3t0fg) ([raw file](https://github.com/zanata/zanata-mt/blob/765963cf6d731a98b2f6ff6140b53e83d5051a04/server/src/main/java/org/zanata/magpie/api/SecurityInterceptor.java#L87)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Do we need to make it `initialPass.equals(token)` to be safe?
</blockquote></details>

changed back to Objects.equal for safety

---

*[server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt, line 58 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3ze2r-ZouHh2_gRFbH:-L4JZYyD5oUak4XGZdSc:b-bs2gxf) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/api/SecurityInterceptorTest.kt#L58)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

That's a shame, because it's kind of bad for test readability. I don't suppose this helps: https://stackoverflow.com/a/22568349/14379
</blockquote></details>

doesn't work for me. Maybe because it conflicts with the junit asserts (auto imported as junit template)? I need to tweak the templates a bit to see if it will work out.

---

*[server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt, line 65 at r3](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L3zY4ZlWCGt8-z5V1uY:-L4JZqJKyG5n7-n-c8jJ:b-896fix) ([raw file](https://github.com/zanata/zanata-mt/blob/61649d5ba8bff9eb61f00434efe752d0a5d7ba07/server/src/test/java/org/zanata/magpie/security/CheckRoleDecisionVoterTest.kt#L65)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

assertj (and thus assertj-core-kotlin) can do things like this:

    assertThat(fellowshipOfTheRing).extracting(TolkienCharacter::getName)
                               .doesNotContain("Sauron", "Elrond");

I think one of the other libraries can do something similar with this feature: https://github.com/kotlintest/kotlintest/blob/master/doc/reference.md#inspectors 

(KotlinTest looks to be quite powerful with its DSL, but as we're used to assertj perhaps we should stick with it for the moment.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by seanf** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-:-L4Jk0hKkk2Z6wGgUeNr:bnfp4nl)*

**Comment by seanf** [comment]
Reviewed 14 of 15 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by huangp** [comment]
@djansen-redhat I've created some end to end tests. e.g. https://github.com/zanata/zanata-mt/pull/53/files#diff-b8f115a3bb724d6512b07ba041c516a2 to test the admin account feature

**Comment by seanf** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

File: server/src/main/resources/db/changelogs/db.changelog-1.1.xml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  <property name="now" value="now()" dbms="postgresql"/>
+
+
+  <changeSet author="pahuang@redhat.com" id="1">
+    <comment>create Account table</comment>
+    <createTable tableName="account">
+      <column name="id" type="SERIAL" autoIncrement="true">
+        <constraints primaryKey="true" />
+      </column>
+      <column name="name" type="varchar(80)" />
```
Comment (djansen-redhat) [inline server/src/main/resources/db/changelogs/db.changelog-1.1.xml]: Mismatch in size constraint to account dto.

File: server/src/main/resources/db/changelogs/db.changelog-1.1.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      <column name="email" type="varchar(100)">
```
Comment (djansen-redhat) [inline server/src/main/resources/db/changelogs/db.changelog-1.1.xml]: We should avoid using arbitrary values for size. Max email size is 254 characters.

File: server/src/main/java/org/zanata/magpie/api/dto/AccountDto.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    @NotNull
+    @Size(min = 1, max = 128)
```
Comment (djansen-redhat) [inline server/src/main/java/org/zanata/magpie/api/dto/AccountDto.java]: Mismatch in size constraint to database value.

File: server/src/main/java/org/zanata/magpie/api/dto/CredentialDto.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public CredentialDto(String username, char[] secret) {
+        this.username = username;
+        this.secret = nullSafeArrayCopy(secret);
+    }
+
+    public CredentialDto() {
+    }
+
+    @NotNull
+    @Size(min = 2, max = 20)
```
Comment (djansen-redhat) [inline server/src/main/java/org/zanata/magpie/api/dto/CredentialDto.java]: Why 20?

File: server/src/main/java/org/zanata/magpie/model/Account.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+@Table(
+        name = "account",
+        uniqueConstraints = @UniqueConstraint(name = "UK_account_email", columnNames = {"email"})
+)
+public class Account extends ModelEntity {
+    private static final long serialVersionUID = -8177299694942674381L;
+    public static final String QUERY_BY_USERNAME = "QUERY_BY_USERNAME";
+
+    @NotNull
+    @Size(min = 1, max = 128)
```
Comment (djansen-redhat) [inline server/src/main/java/org/zanata/magpie/model/Account.java]: Mismatch with database size constraint

File: server/src/main/java/org/zanata/magpie/api/dto/CredentialDto.java (hunk 4 above)
Comment (huangp) [reply server/src/main/java/org/zanata/magpie/api/dto/CredentialDto.java]: Just some arbitrary number. Changed to 128

File: server/src/main/resources/db/changelogs/db.changelog-1.1.xml (hunk 2 above)
Comment (huangp) [reply server/src/main/resources/db/changelogs/db.changelog-1.1.xml]: done

File: server/src/main/resources/db/changelogs/db.changelog-1.1.xml (hunk 1 above)
Comment (huangp) [reply server/src/main/resources/db/changelogs/db.changelog-1.1.xml]: done

File: server/src/main/java/org/zanata/magpie/model/Account.java (hunk 5 above)
Comment (huangp) [reply server/src/main/java/org/zanata/magpie/model/Account.java]: done

**Comment by huangp** [comment]
The new account REST api is for internal use only. Will add to api doc if we want to publish it. So for now I only added the doc to readme. We don't have a comprehensive requirement for all the error handling yet. I implemented something simple to handle database constraint violation. Now it will return status 409 (conflict)

**Comment by djansen-redhat** [comment]
Reviewed 1 of 3 files at r8, 1 of 7 files at r9.
Review status: 49 of 56 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/org/zanata/magpie/exception/DataConstraintViolationException.java, line 1 at r9](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L8-qpw_IiNJy0p99zCL:-L8-qpwaQKcFBorX0ncR:b-24bm1r) ([raw file](https://github.com/zanata/zanata-mt/blob/7ce123d53e2185099dfe5f6bad4f42fc24990191/server/src/main/java/org/zanata/magpie/exception/DataConstraintViolationException.java#L1)):*
> ```Java
> package org.zanata.magpie.exception;
> ```

Licence header

---

*[server/src/main/java/org/zanata/magpie/exception/DataConstraintViolationExceptionMapper.java, line 1 at r9](https://reviewable.io:443/reviews/zanata/zanata-mt/53#-L8-r-UPZEhlTAvguZd3:-L8-r-UPZEhlTAvguZd4:b-24bm1r) ([raw file](https://github.com/zanata/zanata-mt/blob/7ce123d53e2185099dfe5f6bad4f42fc24990191/server/src/main/java/org/zanata/magpie/exception/DataConstraintViolationExceptionMapper.java#L1)):*
> ```Java
> package org.zanata.magpie.exception;
> ```

Licence header

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Comment by djansen-redhat** [comment]
@huangp btw it's 2018 now ;)

**Comment by djansen-redhat** [comment]
Verified. What about coverage?

**Comment by huangp** [comment]
There is not much more we can test in the diff. The overall coverage is still good

**Comment by djansen-redhat** [comment]
Reviewed 1 of 46 files at r1, 1 of 3 files at r8, 4 of 7 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/53)*

**Pull request merged**